### PR TITLE
[mycpp] add data structure builtins for completion

### PR DESCRIPTION
**Pull request by melvinw**
There are a few builtins and data structure methods used throughout the completion code that aren't yet supported in the runtime. This series of commits adds them.

### CODE REVIEW

**Comment by melvinw** [comment]
Slightly easier to review commit-by-commit

File: mycpp/gc_list.h [hunk 1]
```diff
@@ -295,17 +297,15 @@ T List<T>::pop() {
...
   DCHECK(len_ > 0);
-  DCHECK(i == 0);  // only support popping the first item
 
-  T result = index_(0);
+  T result = index_(i);
   len_--;
 
   // Shift everything by one
-  memmove(slab_->items_, slab_->items_ + 1, len_ * sizeof(T));
+  memmove(slab_->items_ + i, slab_->items_ + (i + 1), len_ * sizeof(T));
```
Comment (melvinw) [inline mycpp/gc_list.h:308]: I _think_ this should be fine in general... Perhaps we could add a few extra tests to convince ourselves if unsure

File: mycpp/gc_builtins.cc [hunk 2]
```diff
@@ -291,6 +291,18 @@ bool are_equal(Tuple2<Str*, int>* t1, Tuple2<Str*, int>* t2) {
   return result;
 }
 
+bool are_equal(Tuple2<int, int>* t1, Tuple2<int, int>* t2) {
+  return t1->at0() == t2->at0() && t1->at1() == t2->at1();
+}
+
+bool keys_equal(Tuple2<int, int>* t1, Tuple2<int, int>* t2) {
```
Comment (andychu) [inline mycpp/gc_builtins.cc:298]: I feel like there is some template refactoring to simplify the keys_equal / are_equal stuff, but let's worry about that separately ...

File: mycpp/gc_builtins.h [hunk 3]
```diff
@@ -100,6 +100,13 @@ class OSError : public IOError_OSError {
   }
 };
 
+class SystemExit : public _ExceptionOpaque {
```
Comment (andychu) [inline mycpp/gc_builtins.h:103]: I think this is probably unused given the comment in `core/completion.py`, but it's probably hard to `if mylib.PYTHON` out ...

File: mycpp/gc_list_test.cc [hunk 4]
```diff
@@ -408,6 +408,41 @@ TEST contains_test() {
...
+
+  bool caught = false;
+  try {
+    l->index(3);
+  } catch (ValueError* e) {
+    caught = true;
+  }
+  ASSERT(caught);
+
+  ASSERT_EQ(l->index_(0), 1);
```
Comment (andychu) [inline mycpp/gc_list_test.cc:440]: Generally the style for these types of frameworks is to put the expected value FIRST

If you do `ASSERT_EQ_FMT()` then you'll see `greatest.h` does respect that convention

i.e. it prints a message "Expected: arg1, Actual: arg2"

For these cases it's not a big deal

**Review by andychu — COMMENTED** [review]
Great thanks!  Tests look pretty good

File: mycpp/gc_list_test.cc (hunk 4 above)
Comment (melvinw) [reply mycpp/gc_list_test.cc:440]: Ah whoops. I'm used to gtest which encourages the opposite order (At least in recent years. I think it used to be the other way around). Will send a followup to make these consistent with the rest

**Pull request merged**